cmd/complex/main.go

```diff
@@ -27,26 +28,39 @@ func newSSE() *sse.Server {
 
 	return &sse.Server{
 		Provider: &sse.Joe{Replayer: rp},
-		Logger:   logger{log.New(os.Stderr, "", 0)},
-		OnSession: func(s *sse.Session) (sse.Subscription, bool) {
-			topics := s.Req.URL.Query()["topic"]
+		// If you are using a 3rd party library to generate a per-request logger, this
+		// can just be a simple wrapper over it.
+		Logger: func(r *http.Request) *slog.Logger {
+			l, err := getLogger(r)
+			if err != nil {
+				return nil
+			}
+			return l
+		},
+		OnSession: func(w http.ResponseWriter, r *http.Request) (topics []string, permitted bool) {
+			topics = r.URL.Query()["topic"]
 			for _, topic := range topics {
 				if topic != topicRandomNumbers && topic != topicMetrics {
-					fmt.Fprintf(s.Res, "invalid topic %q; supported are %q, %q", topic, topicRandomNumbers, topicMetrics)
-					s.Res.WriteHeader(http.StatusBadRequest)
-					return sse.Subscription{}, false
+					fmt.Fprintf(
+						w,
+						"invalid topic %q; supported are %q, %q",
+						topic,
+						topicRandomNumbers,
+						topicMetrics,
+					)
```

Comment:
I'd prefer this on a single line
```suggestion
					fmt.Fprintf(w, "invalid topic %q; supported are %q, %q", topic, topicRandomNumbers, topicMetrics)
```